Add support for multi-dimensional array initialization

mikedn opened the pull request:
This extends `impInitializeArrayIntrinsic` to support MD arrays in addition to SD arrays. This includes support for SD arrays which are created via `CORINFO_HELP_NEW_MDARR` when the lower bound is not specified or known to be 0.

The generated code is similar to the code generated for the SD arrays, for example:
```
  call CORINFO_HELP_NEW_MDARR
  mov rdx, 0x202AB822050
  lea rcx, bword ptr [rax+32]
  vmovdqu ymm0, qword ptr [rdx]
  vmovdqu qword ptr [rcx], ymm0
  mov r8d, dword ptr [rdx+16]
  mov dword ptr [rcx+16], r8d
```

Fixes #3839

mmitche commented:
@dotnet-bot test this please

mikedn commented:
I've done some assembly diffs (using crossgen and COMPLUS_NgenDisasm=*) on mscorlib, corefx and the Roslyn compilers. There are a handful of differences (most in mscorlib and Microsoft.CodeAnalysis.CSharp/VisualBasic) all of the form: 

```
        488B0D00000000       mov      rcx, qword ptr [(reloc)]
        BA02000000           mov      edx, 2
        41B810000000         mov      r8d, 16
        41B910000000         mov      r9d, 16
        FF1500000000         call     [CORINFO_HELP_NEW_MDARR]
        488BF0               mov      rsi, rax
-       488B0D00000000       mov      rcx, qword ptr [(reloc)]
-       FF1500000000         call     [CORINFO_HELP_FIELDDESC_TO_STUBRUNTIMEFIELD]
-       488BD0               mov      rdx, rax
-       488BCE               mov      rcx, rsi
-       FF1500000000         call     [System.Runtime.CompilerServices.RuntimeHelpers:InitializeArray(ref,struct)]
+       488D1500000000       lea      rdx, [(reloc)]
+       488D4E20             lea      rcx, bword ptr [rsi+32]
+       41B800010000         mov      r8d, 256
+       FF1500000000         call     [CORINFO_HELP_MEMCPY]
        FF1500000000         call     [CORINFO_HELP_READYTORUN_STATIC_BASE]
        488D8888040000       lea      rcx, bword ptr [rax+0488H]
        488BD6               mov      rdx, rsi
        FF1500000000         call     [CORINFO_HELP_ASSIGN_REF]
        90                   nop      
```

jkotas commented:
cc @dotnet/jit-contrib @MichalStrehovsky

jkotas commented:
@mikedn I am sorry this has been sitting here for so long without anybody taking a look. The change looks good to me. Thanks for adding all the test cases!

@dotnet/jit-contrib Any other comments? (I am sure that the JIT coding conventions are not followed in there somewhere ;-)

As we you have pointed out, this does not work for the non-vararg helper as is. Would you mind taking a look at changing the pattern matching to recognize the non-vararg case once my change with the non-vararg helper goes in? I do not think it would be too bad. We should just worry about the non-vararg helper because of it is what we will switch to eventually everywhere, and it is what is needed to make it work well for CoreRT - that motivated this issue to start with.

mikedn commented:
>  Would you mind taking a look at changing the pattern matching to recognize the non-vararg case once my change with the non-vararg helper goes in?

Of course I don't mind, I'll take a look in a day or two. I hope it's possible to recognize the new pattern but if it's not then an alternative may exist: recognize during the initial import (should be easier) and somehow attach the necessary information (bound lengths) to the call node.

CarolEidt commented:
@mikedn Sorry that nobody on the jit team has looked at this yet. I'm looking now. @jkotas - so far it looks like the new code is way more aligned with the coding conventions that the old :-)

CarolEidt commented on src/jit/importer.cpp:
```diff
@@ -2663,23 +2663,30 @@ BOOL            Compiler::impLocAllocOnStack()
+//    sig - The InitializeArray signature.
+//
+// Return Value:
+//    A pointer to the newly create GT_COPYBLK node if the replacement succeeds or
```
Typo: create -> created

CarolEidt commented on src/jit/importer.cpp:
```diff
@@ -2663,23 +2663,30 @@ BOOL            Compiler::impLocAllocOnStack()
+//      call InitializeArray
+//    The lower bounds need not be constant except when the array rank is 1.
+//    The function recognizes all kinds of arrays thus enabling a small runtime 
+//    such as CoreRT to skip providing an implementation for InitializeArray.
```
Since the intent is for this to allow a runtime to not provide an implementation for InitializeArray, would it make sense to have an ifdef for those runtimes so that instead of returning nullptr, this would assert? Since, if there's no implementation, it would presumably be an error.

jkotas replied:
This error handling is in impIntrinsic. There is a flag passed over JIT/EE interface that tells the JIT whether the intrinsic has to be expanded unconditionally, or whether it is just an optimization.

CarolEidt replied:
That makes sense, thanks.

CarolEidt commented on src/jit/importer.cpp:
```diff
@@ -2773,57 +2783,177 @@ GenTreePtr      Compiler::impInitializeArrayIntrinsic(CORINFO_SIG_INFO * sig)
+                    if (lowerBoundNode->OperGet() != GT_CNS_INT)
+                        return nullptr;
+
+                    if (lowerBoundNode->AsIntCon()->IconValue() == 0)
```
Minor suggestion: You could simply check `if (lowerBoundNode->IsZero())` here, and omit the check for non-constant, which will be taken care of below - makes the code a little simpler to understand.

CarolEidt commented:
Looks good - nicely done :+1:

sivarv commented on src/jit/importer.cpp:
```diff
@@ -2773,57 +2783,177 @@ GenTreePtr      Compiler::impInitializeArrayIntrinsic(CORINFO_SIG_INFO * sig)
+
+            GenTree* lengthNode = arg->Current();
+
+            if (lengthNode->OperGet() != GT_CNS_INT)
```
We can use GenTree::IsCnsIntOrI()

e.g.

```
if (!lengthNode->IsCnsIntOrI())
{
    return nullptr
}
```

sivarv commented on src/jit/importer.cpp:
```diff
@@ -2773,57 +2783,177 @@ GenTreePtr      Compiler::impInitializeArrayIntrinsic(CORINFO_SIG_INFO * sig)
+
+        if ((beginArgs == nullptr) ||
+            (numArgsArg == nullptr) ||
+            (numArgsArg->Current()->OperGet() != GT_CNS_INT) ||
```
Above condition could be written using IsCnIntOrI()

!numArgsArg->Current()->IsCnsIntOrI()

sivarv commented:
Looks good.

mikedn commented:
I have experimented with adding support for `CORINFO_HELP_NEW_MDARR_NONVARARG` and it seems that it manageable if we're comfortable with looking only at assignments and their RHS.

My Internet access is kind of limited these days and I can't to do a build (too much to download to get build tools). I hope to be back in a couple of days.

mikedn commented:
Rebased, addressed feedback and switched to the non-vararg helper (the vararg helper is no longer recognized).

I've ended up pattern matching the entire "args" tree but most of the matching is done only in asserts. Since the IR shape does not depend on the input IL it seems unnecessary to use something stronger than an assert.

jkotas commented:
LGTM. Thanks a lot!
